tests/manage/configuration/migrations/test_migrations.py

```diff
@@ -25,108 +26,70 @@ class CircularMigration(Initial):
-        config_initial = self.load_config_sample('input/5.json')
-        config_reference = self.load_config_sample('reference/6.json')
-        config_migrated = upgrade_config(config_initial, 6)
-        self.assertEqual(config_migrated, config_reference)
-
-    def test_6_5_migration(self):
-        """Test the step between config versions 6 and 5."""
-        config_initial = self.load_config_sample('reference/6.json')
-        config_reference = self.load_config_sample('input/5.json')
-        config_migrated = downgrade_config(config_initial, 5)
-        self.assertEqual(config_migrated, config_reference)
-
-    def test_6_7_migration(self):
-        """Test the step between config versions 6 and 7."""
-        config_initial = self.load_config_sample('input/6.json')
-        config_reference = self.load_config_sample('reference/7.json')
-        config_migrated = upgrade_config(config_initial, 7)
-        self.assertEqual(config_migrated, config_reference)
+    return _factory
+
+
+def test_upgrade_path_fail(load_config_sample):
+    """Test failure when no upgrade path is available."""
+    config_initial = load_config_sample('reference/5.json')
+    # target lower than initial
+    with pytest.raises(ConfigurationError):
+        upgrade_config(copy.deepcopy(config_initial), 1)
+    # no migration available
+    with pytest.raises(ConfigurationError):
+        upgrade_config(copy.deepcopy(config_initial), 100)
+    # circular dependency
+    with pytest.raises(ConfigurationError):
+        upgrade_config(config_initial, 6, migrations=[CircularMigration])
+
+
+def test_downgrade_path_fail(load_config_sample):
+    """Test failure when no downgrade path is available."""
+    config_initial = load_config_sample('reference/5.json')
+    # target higher than initial
+    with pytest.raises(ConfigurationError):
+        downgrade_config(copy.deepcopy(config_initial), 6)
+    # no migration available
+    with pytest.raises(ConfigurationError):
+        downgrade_config(copy.deepcopy(config_initial), -1)
+    # circular dependency
+    with pytest.raises(ConfigurationError):
+        downgrade_config(config_initial, 4, migrations=[CircularMigration])
+
+
+def test_migrate_full(load_config_sample, monkeypatch):
+    """Test the full config migration."""
+    config_initial = load_config_sample('input/0.json')
+    monkeypatch.setattr(uuid, 'uuid4', lambda: uuid.UUID(hex='0' * 32))
+    config_migrated = check_and_migrate_config(config_initial)
+    config_reference = load_config_sample('reference/final.json')
+    assert config_migrated == config_reference
+
+
+@pytest.mark.parametrize('initial, target', (
+    (0, 1),
+    (1, 2),
+    (2, 3),
+    (3, 4),
+    (4, 5),
+    (5, 6),
+    (6, 7),
+    (7, 8),
+))
+def test_migrate_individual(load_config_sample, initial, target, monkeypatch):
+    """Test the individual config migrations."""
+    monkeypatch.setattr(uuid, 'uuid4', lambda: uuid.UUID(hex='0' * 32))
```

Comment:
Could indeed put in a `if target ==2:` conditional.

Reply:
Since these are all single step updates that would be fine, yes, but I think strictly speaking the condition is `initial < 2` no?

Reply:
I don't think so, the UUID is only generated when migrating up to v2, so we only need it for that step.

Reply:
Mmm, ok, I think I see your point, if we were doing 0 to 1 then `initial < 2` is true but there is no `UUID` generated, so you are right my condition is not correct. What I was pointing out originally is that if you did several migrations (so for example, from 1 to 3) you would still generate a new `UUID` even if you didn't have `target == 2`. So I guess the minimal strict rule would then be `initial < 2 and target >= 2`, right?

(again, this is just to check that I correctly understand the logic, since these are all single step you can leave `target == 2`)

Reply:
Yes, you are right, the monkeypatch should be applied for any migration path that contains `1 -> 2`